File: Packages/Pets/Sources/Pets/Entity/PetEntity.swift
import Combine
import Schwifty
import SwiftUI
import Yage

public protocol PetsSettings {
    var gravityEnabled: Bool { get }
    var petSize: CGFloat { get }
    var speciesOnStage: CurrentValueSubject<[Species], Never> { get }
    var speedMultiplier: CGFloat { get }
    var ufoAbductionSchedule: String { get }
    func remove(species: Species)
}

open class PetEntity: Entity {
    public static var assetsProvider: AssetsProvider?
    public let settings: PetsSettings

    public init(of species: Species, in world: World, settings: PetsSettings) {
        self.settings = settings
        super.init(
            species: species,
            id: PetEntity.id(for: species),
            frame: PetEntity.initialFrame(in: world.bounds, size: settings.petSize),
            in: world
        )
        loadProperties()
    }

    func loadProperties() {
        resetSpeed()
        setInitialPosition()
        setInitialDirection()
        setGravity()
    }

    override open func set(state: EntityState) {
        super.set(state: state)
        if case .move = state { resetSpeed() }
    }

    public func resetSpeed() {
        speed = PetEntity.speed(
            for: species,
            size: frame.width,
            settings: settings.speedMultiplier
        )
    }

    private func setInitialPosition() {
        let randomX = CGFloat.random(in: 0.1 ..< 0.75) * worldBounds.width 
        let randomY: CGFloat

        if capability(for: WallCrawler.self) != nil {
            randomY = worldBounds.height - frame.height
        } else {
            randomY = CGFloat.random(in: 0.1 ..< 0.5) * worldBounds.height
        }
        frame.origin = CGPoint(x: randomX, y: randomY)
    }

    private func setInitialDirection() {
        direction = .init(dx: 1, dy: 0)
    }

    public var supportsGravity: Bool {
        capability(for: WallCrawler.self) == nil
    }

    public func setGravity() {
        setGravity(enabled: settings.gravityEnabled && supportsGravity)
    }
}

File: Sources/windows/onscreen/pet_entity.py
import random
from config.config import Config
from di import *
from yage.capabilities import Gravity
from yage.models import *
from yage.utils.geometry import *

class PetEntity(Entity):
    def __init__(self, species, world):
        self.config = Dependencies.instance(Config)
        pet_size = self.config.pet_size.value
        gravity_enabled = self.config.gravity_enabled.value

        super().__init__(
            species, 
            PetEntity.next_id(species), 
            Rect(0, 0, pet_size, pet_size),
            world
        )
        self.reset_speed()
        self.place_in_random_position()
        self._set_initial_direction()
        self.set_gravity_enabled(gravity_enabled)
    
    def reset_speed(self):
        config = self.config.speed_multiplier.value
        size = self.frame.size.width / 75.0
        self.speed = config * 30.0 * self.species.speed * size

    def place_in_random_position(self):
        random_x = random.uniform(0.1, 0.75) * self.world_bounds.width
        self.frame.origin = Point(random_x, self.world_bounds.height+5)

    def _set_initial_direction(self):
        self.direction = Vector(1, 0)

    def set_gravity_enabled(self, enabled: bool):
        gravity = self.capability(Gravity)
        if not gravity: return
        gravity.is_enabled = enabled
        if not enabled:
            if self.direction.dy > 0:
                self.direction = Vector(1, 0)
            self.state = EntityState.MOVE

    @classmethod
    def next_id(cls, species):
        if not hasattr(cls, 'incremental_id'): cls.incremental_id = 0
        cls.incremental_id += 1
        return f'{species.id}-{cls.incremental_id}'
